File: problems/3176_find_the_maximum_length_of_a_good_subsequence_i/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int maximumLength(vector<int>& nums, int k) {
        int n = nums.size(), res = 1;
        vector<vector<int>> dp(n, vector<int>(k+1, 1));
        for (int i = 0; i < n; i++)
            for (int j = 0; j <= k; j++) {
                for (int p = 0; p < i; p++) {
                    if (nums[p]==nums[i]) dp[i][j]=max(dp[i][j],dp[p][j]+1);
                    else if (j>0) dp[i][j]=max(dp[i][j],dp[p][j-1]+1);
                }
                res = max(res, dp[i][j]);
            }
        return res;
    }
};

File: problems/3176_find_the_maximum_length_of_a_good_subsequence_i/solution.py
from typing import List


class Solution:
    def maximumLength(self, nums: List[int], k: int) -> int:
        n = len(nums)
        dp = [[0] * (k + 1) for _ in range(n)]
        res = 1
        for i in range(n):
            for j in range(k + 1):
                dp[i][j] = 1
                for p in range(i):
                    if nums[p] == nums[i]:
                        dp[i][j] = max(dp[i][j], dp[p][j] + 1)
                    elif j > 0:
                        dp[i][j] = max(dp[i][j], dp[p][j - 1] + 1)
                res = max(res, dp[i][j])
        return res
